mycpp/control_flow_pass.py

```diff
@@ -24,19 +25,98 @@ class UnsupportedException(Exception):
 
 class Build(ExpressionVisitor[T], StatementVisitor[None]):
 
-    def __init__(self, types: Dict[Expression, Type]):
+    def __init__(self, types: Dict[Expression, Type], virtual, local_vars):
 
         self.types = types
-        self.cfgs = {}
+        self.cfgs = collections.defaultdict(pass_state.ControlFlowGraph)
         self.current_statement_id = None
+        self.current_class_name = None
         self.current_func_node = None
         self.loop_stack = []
+        self.virtual = virtual
+        self.local_vars = local_vars
+        self.imported_names = set()  # MemberExpr -> module::Foo() or self->foo
+
+        # HACK for conditional import inside mylib.PYTHON
+        # in core/shell.py
+        self.imported_names.add('help_meta')
```

Comment:
We shouldn't duplicate this imported_names structure here, and also in cppgen_pass.py

Because it is heuristic, and the two heuristics can get out of sync

Generally we want to compute things in one pass, and then pass something more structured and rigid on to future passes

---

I just did some of that to fix rooting ... in the DECL pass we now compute `ctx_member_vars`, and then we pass that info onto the IMPL pass

(and eventually we are going to get rid of the self.decl and self.forward_decl booleans and factor out all the duplication)

Reply:
Ah sure. I thought this logic was simple enough and unlikely to change, but it's better to be safe. I'll pass this in.